### PR TITLE
[Feature] support partial update by column, handle write

### PR DESCRIPTION
**Pull request by luohaha**
Signed-off-by: luohaha <18810541851@163.com>

## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #20436

## Problem Summary(Required) ：
This PR is the third part of partial update column mode feature. Add `HorizontalUpdateRowsetWriter` to write `.upt` files and apply logic about partial update.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

File: be/src/storage/rowset/rowset_writer.h [hunk 1]
```diff
@@ -229,6 +238,23 @@ class HorizontalRowsetWriter final : public RowsetWriter {
     std::unique_ptr<VerticalRowsetWriter> _vertical_rowset_writer;
 };
 
+// used for column mode partial update
+class HorizontalUpdateRowsetWriter final : public RowsetWriter {
```
Comment (imay) [inline be/src/storage/rowset/rowset_writer.h:242]: How about writing it in a new file?

File: be/src/storage/delta_writer.h [hunk 2]
```diff
@@ -64,6 +64,7 @@ struct DeltaWriterOptions {
     ReplicaState replica_state;
     bool miss_auto_increment_column = false;
     bool abort_delete = false;
+    PartialUpdateMode partial_update_mode = PartialUpdateMode::ROW_MODE;
```
Comment (chaoyli) [inline be/src/storage/delta_writer.h:67]: 1. It's better to determine the ROW_MODE or COLUMN_MODE when writing.
2. If using the `partial_update_mode` in the stream load, the user can not get an idea of how many segments will be included in one ingestion.

Comment (luohaha) [reply be/src/storage/delta_writer.h:67]: After discussion, keep `partial_update_mode`, and add `PartialUpdateMode::AUTO_MODE`, make it a session variable. Using `partial_update_mode=PartialUpdateMode::AUTO_MODE` as default value, SR will decide  ROW_MODE or COLUMN_MODE when writing.

File: be/src/storage/rowset/rowset_writer.h (hunk 1 above)
Comment (luohaha) [reply be/src/storage/rowset/rowset_writer.h]: Done

File: be/src/storage/rowset/horizontal_update_rowset_writer.cpp [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+HorizontalUpdateRowsetWriter::HorizontalUpdateRowsetWriter(const RowsetWriterContext& context)
+        : RowsetWriter(context), _update_file_writer(nullptr) {}
+
+HorizontalUpdateRowsetWriter::~HorizontalUpdateRowsetWriter() {
+    if (!_already_built) {           // abnormal exit, remove all files generated
+        _update_file_writer.reset(); // ensure all files are closed
+        for (auto i = 0; i < _num_uptfile; ++i) {
+            auto path = Rowset::segment_upt_file_path(_context.rowset_path_prefix, _context.rowset_id, i);
+            auto st = _fs->delete_file(path);
+            LOG_IF(WARNING, !(st.ok() || st.is_not_found()))
```
Comment (chaoyli) [inline be/src/storage/rowset/horizontal_update_rowset_writer.cpp:33]: add the filename to the `WARNING` log.

File: be/src/storage/rowset/rowset.cpp [hunk 4]
```diff
@@ -294,7 +294,7 @@ Status Rowset::_remove_delta_column_group_files(std::shared_ptr<FileSystem> fs)
             for (const auto& dcg : list) {
                 auto st = fs->delete_file(dcg->column_file());
                 if (st.ok() || st.is_not_found()) {
-                    LOG(INFO) << "Deleting delta column group's file: " << dcg->debug_string() << " st: " << st;
+                    VLOG(1) << "Deleting delta column group's file: " << dcg->debug_string() << " st: " << st;
```
Comment (chaoyli) [inline be/src/storage/rowset/rowset.cpp:297]: Add the filename to the log

File: be/src/storage/rowset/rowset_writer.cpp [hunk 5]
```diff
@@ -209,104 +211,161 @@ StatusOr<RowsetSharedPtr> RowsetWriter::build() {
...
+            return io::io_error(wfile->filename(), errno);
         }
+        remaining_bytes -= written_bytes;
+    }
+    if (remaining_bytes != 0) {
+        return Status::InternalError(fmt::format("segment {} write size {} not equal expected size {}",
+                                                 wfile->filename(), segment_pb.data_size() - remaining_bytes,
+                                                 segment_pb.data_size()));
+    }
+    if (config::sync_tablet_meta) {
```
Comment (chaoyli) [inline be/src/storage/rowset/rowset_writer.cpp:241]: Can we flush only once for one ingestion?

File: be/src/storage/rowset/rowset_writer.cpp [hunk 6]
```diff
@@ -209,104 +211,161 @@ StatusOr<RowsetSharedPtr> RowsetWriter::build() {
...
+    // use MUST_CREATE make sure atomic
+    ASSIGN_OR_RETURN(auto wfile, _fs->new_writable_file(path));
+
+    // 2. flush delete file
+    auto writer = std::make_unique<SegmentFileWriter>(wfile.get());
+
+    butil::IOBuf delete_data;
+    int64_t remaining_bytes = data.cutn(&delete_data, segment_pb.delete_data_size());
+    if (remaining_bytes != segment_pb.delete_data_size()) {
+        return Status::InternalError(fmt::format("segment {} delete size {} not equal attachment size {}",
```
Comment (chaoyli) [inline be/src/storage/rowset/rowset_writer.cpp:279]: How about the segment_pb size? Maybe it is too large to place the debug string in the returned status?

File: be/src/storage/rowset/rowset.cpp (hunk 4 above)
Comment (luohaha) [reply be/src/storage/rowset/rowset.cpp:297]: The filename is included in `debug_string()`

File: be/src/storage/rowset/horizontal_update_rowset_writer.cpp (hunk 3 above)
Comment (luohaha) [reply be/src/storage/rowset/horizontal_update_rowset_writer.cpp:33]: It already added. Check the next line.

File: be/src/storage/rowset/rowset_writer.cpp (hunk 6 above)
Comment (luohaha) [reply be/src/storage/rowset/rowset_writer.cpp:279]: change to segment's filename.

File: be/src/storage/tablet_updates.cpp [hunk 7]
```diff
@@ -1360,6 +1460,40 @@ Status TabletUpdates::_do_compaction(std::unique_ptr<CompactionInfo>* pinfo) {
...
+                need_cancel = true;
+            }
+        }
+
+        if (need_cancel) {
+            LOG(WARNING) << strings::Substitute("compaction conflict with partial update failed, tabletid:$0 ver:$1-$2",
+                                                _tablet.tablet_id(), info->start_version.major(),
+                                                (*i)->version.major());
+            _compaction_state.reset();
+            return Status::Cancelled("compaction conflict with partial update");
```
[the remark this answers is not in the document]
Comment (luohaha) [reply be/src/storage/tablet_updates.cpp]: Done

File: be/src/storage/rowset/rowset_writer.cpp (hunk 5 above)
Comment (luohaha) [reply be/src/storage/rowset/rowset_writer.cpp:241]: This `sync` will be called only once for each segment. It is just ok?

Comment (chaoyli) [reply be/src/storage/rowset/rowset_writer.cpp:241]: OK

**Review by decster — APPROVED** [review]

**Review by chaoyli — APPROVED** [review]

**Pull request merged**